### PR TITLE
Introduce to weak model

### PR DESCRIPTION
**Pull request by KennyDizi**
### **PR Type**
Enhancement, Documentation

___

### **Description**
- Introduced a new `WEAK` model type, replacing the previous `TURBO` model.
- Updated various functions to utilize the `WEAK` model as the default for specific flows.
- Revised documentation to guide users on configuring the new `WEAK` model.
- Adjusted configuration files to reflect the new model settings.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>10 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>pr_processing.py</strong><dd><code>Update model handling to use weak model by default</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/algo/pr_processing.py

<li>Changed default model type from <code>REGULAR</code> to <code>WEAK</code> in <br><code>retry_with_fallback_models</code>.<br> <li> Updated <code>_get_all_models</code> to handle <code>WEAK</code> model type.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1387/files#diff-c2d64de584ef3cb05c9007d72137f6c7ce04c9bc23ce1931760af3a568edb04e">+4/-4</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>utils.py</strong><dd><code>Modify ModelType Enum to include weak model</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/algo/utils.py

- Removed `TURBO` model type.
- Added `WEAK` model type.

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1387/files#diff-6b9df72d53c6f0d89fb142c210238a276c0782305e0024d16fbfcaf72c2e2b53">+1/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>utils.py</strong><dd><code>Update model configuration for weak model</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/git_providers/utils.py

<li>Updated configuration to set <code>model_weak</code> instead of <code>model_turbo</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1387/files#diff-314199df3610cf35a472e952294faf0f5fbb516a2bd217e891a510fd1a698bca">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>pr_code_suggestions.py</strong><dd><code>Use weak model for code suggestions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/tools/pr_code_suggestions.py

<li>Changed model type used in <code>retry_with_fallback_models</code> from <code>REGULAR</code> to <br><code>WEAK</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1387/files#diff-b57ba775e741d6f80bc4f8154b71330c011dae0ac43f3d0197e785b3e6b7117b">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>pr_description.py</strong><dd><code>Switch to weak model for PR description generation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/tools/pr_description.py

- Updated model type in `retry_with_fallback_models` to `WEAK`.

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1387/files#diff-66130451adce278a098a5770e2db52b12050205fcd4d80c13f813615d4449abb">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>pr_help_message.py</strong><dd><code>Adjust help message generation to use weak model</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/tools/pr_help_message.py

- Changed model type in `retry_with_fallback_models` to `WEAK`.

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1387/files#diff-6f30977025b20a27640fc33cfe2db79160753bd30c7879f60f2b5894d86f78de">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>pr_line_questions.py</strong><dd><code>Modify line question handling to use weak model</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/tools/pr_line_questions.py

- Updated model type in `retry_with_fallback_models` to `WEAK`.

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1387/files#diff-7fc24c7ebf69233b72a5f56d003cc266ca0a952ecc7cf2e75123d0a1ab87d525">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>pr_questions.py</strong><dd><code>Use weak model for question handling</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/tools/pr_questions.py

- Changed model type in `retry_with_fallback_models` to `WEAK`.

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1387/files#diff-71970fafa37a18dfb27eb7b47754bfa7841e14686b181445298f914e03fc31aa">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>pr_reviewer.py</strong><dd><code>Ensure reviewer uses regular model</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/tools/pr_reviewer.py

- Updated model type in `retry_with_fallback_models` to `REGULAR`.

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1387/files#diff-8e265068e189a06852605cc694b03e92b523b4f8162077a2f3455ced4cdad8dc">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>pr_update_changelog.py</strong><dd><code>Update changelog generation to use weak model</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/tools/pr_update_changelog.py

- Changed model type in `retry_with_fallback_models` to `WEAK`.

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1387/files#diff-76d8133da16132fa69d308bf1e5edb4fc7de18343a4d99760960416c54c79087">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>changing_a_model.md</strong><dd><code>Revise usage guide for weak model configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

docs/docs/usage-guide/changing_a_model.md

<li>Updated documentation to reflect changes from <code>model_turbo</code> to <br><code>model_weak</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1387/files#diff-7904af6a8722d533e7de626761ae6c1d9554069ed02634c3b719d2e7f8b39f4d">+11/-11</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Configuration changes</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>configuration.toml</strong><dd><code>Update configuration for weak model settings</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/settings/configuration.toml

<li>Set <code>model_weak</code> to a specific model version.<br> <li> Adjusted default model configuration.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1387/files#diff-66cfda5143e484ee53ecf7aa0df7dca8ad0b181256f4b0675905db35bcbbae78">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

**Comment by KennyDizi** [comment]
Preparing PR description...

**Comment by KennyDizi** [comment]
**[PR Description](https://github.com/Codium-ai/pr-agent/pull/1387)** updated to latest commit (https://github.com/Codium-ai/pr-agent/commit/e3d779c30dafd804d3d1617ab3e60f4776d8e6f1)

File: pr_agent/algo/pr_processing.py [hunk 1]
```diff
@@ -333,7 +333,7 @@ def generate_full_patch(convert_hunks_to_line_numbers, file_dict, max_tokens_mod
     return total_tokens, patches, remaining_files_list_new, files_in_patch_list
 
 
-async def retry_with_fallback_models(f: Callable, model_type: ModelType = ModelType.REGULAR):
+async def retry_with_fallback_models(f: Callable, model_type: ModelType = ModelType.WEAK):
```
Comment (mrT23) [inline pr_agent/algo/pr_processing.py:336]: the default model should be the regular model

File: pr_agent/algo/pr_processing.py [hunk 2]
```diff
@@ -353,9 +353,9 @@ async def retry_with_fallback_models(f: Callable, model_type: ModelType = ModelT
                 raise Exception(f"Failed to generate prediction with any model of {all_models}")
 
 
-def _get_all_models(model_type: ModelType = ModelType.REGULAR) -> List[str]:
-    if model_type == ModelType.TURBO:
-        model = get_settings().config.model_turbo
+def _get_all_models(model_type: ModelType = ModelType.WEAK) -> List[str]:
+    if model_type == ModelType.WEAK:
+        model = get_settings().config.model_weak
```
Comment (mrT23) [inline pr_agent/algo/pr_processing.py:358]: the default model should be the regular model.
Also, we talked about utilizing this only if two conditions are met:
- a weak model is defined
- the tool asks for a weak model.

So something like:

```python
if get_settings().config.get('model_weak') and model_type == ModelType.WEAK:
  model = get_settings().config.model_weak
else:
  model = get_settings().config.model
```

Comment (mrT23) [reply pr_agent/algo/pr_processing.py:358]: note that PR-Agent gave a similar suggestion :)

image

Comment (KennyDizi) [reply pr_agent/algo/pr_processing.py:358]: Good catch, I will adjust the code accordingly.

**Comment by KennyDizi** [comment]
@mrT23 I've updated the PR according to your comments, pls take a look at it again.

**Pull request merged**